pkg/messageformat/extract.go

```diff
@@ -0,0 +1,193 @@
+package messageformat
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+)
+
+// specialChars is a list of characters that need to be escaped in the node.Text,
+// because they are either syntax characters or reserved keywords.
+var specialChars = map[rune]struct{}{
+	// Syntax characters
+	'{':  {},
+	'}':  {},
+	'\\': {},
+	'|':  {},
+	// Reserved keywords (Future syntax characters)
+	'!': {},
+	'@': {},
+	'#': {},
+	'%': {},
+	'*': {},
+	'<': {},
+	'>': {},
+	'/': {},
+	'?': {},
+	'~': {},
+}
+
+const maxOptionCount = 3 // maximum size of []Node.Option, when parsing a placeholder
+
+type placeholderFormat struct {
+	re        *regexp.Regexp
+	nodeExprF func(string, []int) NodeExpr
+}
+
+// printfVerbs maps printf verbs to their corresponding type.
+var printfVerbs = map[string]string{
+	"s": "string",
+	"d": "int",
+	"f": "float",
+}
+
+// placeholderFormats maps placeholder formats to their corresponding regex and NodeExpr factory function.
+var placeholderFormats = map[string]placeholderFormat{
+	"pythonVar": { // hello %(var)s | hello %(var)d
+		re:        regexp.MustCompile(`%\((\w+)\)([sd])`),
+		nodeExprF: createNodeExpr("pythonVar"),
+	},
+	"printf": { // hello %s | hello %d
+		re:        regexp.MustCompile(`%(s|d)`),
+		nodeExprF: createNodeExpr("printf"),
+	},
+	"bracketVar": { // hello {var} | hello {0}
+		re:        regexp.MustCompile(`\{(\w+)\}`),
+		nodeExprF: createNodeExpr("bracketVar"),
+	},
+	"emptyBracket": { // hello {}
+		re:        regexp.MustCompile(`\{\}`),
+		nodeExprF: createNodeExpr("emptyBracket"),
+	},
+}
+
+// messageToAST converts the given message to an MF2 AST.
+func (pf *placeholderFormat) messageToAST(message string) AST {
```

Comment:
why not pass in *placeholderFormat as an argument?

Reply:
just thought, that it belongs as a method for struct, more than an standalone function 🤷

Reply:
@klavinsjanis , there are two benefits of using pointer receivers:
* make changes to the receiver instance (as you said)
* avoid copying large instances by value. E.g. if the receiver instance uses 1KB of memory, passing by value is very expensive, passing by pointer makes it cheaper.

Reply:
I know about pointers, but just was wondering why method is used instead of function.